src/packet/pacer.rs

```diff
@@ -116,6 +116,28 @@ pub struct QueueSnapshot {
     pub last_emitted: Option<Instant>,
     /// Time the first unsent packet has spent in the queue.
     pub first_unsent: Option<Instant>,
+    /// The priority of the most important packet in the queue.
+    pub priority: QueuePriority,
+}
```

Comment:
At the very least we need to move `QueuePriority` up on `QueueState`. The snapshot should just be a statement of fact "this is how the buffer looks", without any judgement calls on priority. That can be introduced on the next level up.

However overall I'm not quite feeling this `QueuePriority` – functionally it's hardly used in Pacer and it complicates the code. It creates a confusion because the _actual_ prio is what `poll_packet` does, i.e. resends first, regular second and padding last.

If we need an indication of which mid contains the most regular and resends (not padding), we could add a secondary counter into `QueueState`, so that we have `padding_count_padding` and `padding_count_regular`.

Reply:
> At the very least we need to move QueuePriority up on QueueState. The snapshot should just be a statement of fact "this is how the buffer looks", without any judgement calls on priority. That can be introduced on the next level up.

We only return a single `QueueState` per `MediaInner`, but that is summarising many internal queues. You are right that it could move to `QueueState`, but it's not a fixed field still, it needs to be calculated based on the actual state of the queues.

> However overall I'm not quite feeling this QueuePriority – functionally it's hardly used in Pacer and it complicates the code. It creates a confusion because the actual prio is what poll_packet does, i.e. resends first, regular second and padding last.

The priority is always media first(whether resend or new) then padding. The pacer must know the state of each queue to avoid requesting padding from one mid when another contains queued media.  

> If we need an indication of which mid contains the most regular and resends (not padding), we could add a secondary counter into QueueState, so that we have padding_count_padding and padding_count_regular.

This is not what we need, the pacer needs to know among all the mids which have media vs those that only contain padding.

Reply:
Ok. Merge away, but I might circle back to this.

Reply:
Sound good to me, the invariant to maintain is:

If we have mid 1 with only padding queued and mid 2 with media and padding queued, pacer must never emit `PollOutcome::PollQueue("1".into())`.